src/txreconciliation.h

```diff
@@ -0,0 +1,585 @@
+            if (local_tx != m_local_short_id_mapping.end()) {
+                remote_missing.push_back(local_tx->second);
+            }
+        }
+        return remote_missing;
+    }
+
+    /**
+     * TODO document
+     */
+    void FinalizeIncomingReconciliation()
+    {
+        assert(m_requestor);
+        ClearState();
+    }
+
+    /**
+     * TODO document
+     */
+    void FinalizeOutgoingReconciliation(bool clear_local_set, double updated_q)
+    {
+        assert(m_responder);
+        m_local_q = updated_q;
+        if (clear_local_set) m_local_set.clear();
+        ClearState();
+    }
+
+    /**
+     * When during reconciliation we find a set difference successfully (by combining sketches),
+     * we want to find which transactions are missing on our and on their side.
+     * For those missing on our side, we may only find short IDs.
+     */
+    std::pair<std::vector<uint32_t>, std::vector<uint256>> GetRelevantIDsFromShortIDs(const std::vector<uint64_t>& diff) const
+    {
+        std::vector<uint32_t> local_missing;
+        std::vector<uint256> remote_missing;
+        for (const auto& diff_short_id: diff) {
+            const auto local_tx = m_local_short_id_mapping.find(diff_short_id);
+            if (local_tx != m_local_short_id_mapping.end()) {
+                remote_missing.push_back(local_tx->second);
+            } else {
+                local_missing.push_back(diff_short_id);
+            }
+        }
+        return std::make_pair(local_missing, remote_missing);
+    }
+
+    /**
+     * TODO document
+     */
+    void PrepareForExtensionRequest(uint16_t sketch_capacity)
+    {
+        // Be ready to respond to extension request, to compute the extended sketch over
+        // the same initial set (without transactions received during the reconciliation).
+        // Allow to store new transactions separately in the original set.
+        assert(m_requestor);
+        m_capacity_snapshot = sketch_capacity;
+        m_local_set_snapshot = m_local_set;
+        m_local_set.clear();
+    }
+
+    /**
+     * TODO document
+     */
+    void PrepareForExtensionResponse(uint16_t sketch_capacity, const std::vector<uint8_t>& remote_sketch)
+    {
+        assert(m_responder);
+        m_capacity_snapshot = sketch_capacity;
+        m_remote_sketch_snapshot = remote_sketch;
+        m_local_set_snapshot = m_local_set;
+        m_local_set.clear();
+    }
+};
+
+/**
+ * Used to track reconciliations across all peers.
+ */
+class TxReconciliationTracker {
+    /**
+     * Salt used to compute short IDs during transaction reconciliation.
+     * Salt is generated randomly per-connection to prevent linking of
+     * connections belonging to the same physical node.
+     * Also, salts should be different per-connection to prevent halting
+     * of relay of particular transactions due to collisions in short IDs.
+     */
+    Mutex m_local_salts_mutex;
```

Comment:
Why multiple mutexes here? I think one mutex for the entire class would suffice.